lab/src/raft/leader_election.go

```diff
@@ -10,21 +10,88 @@ func (rf *Raft) toCandidate() {
 	rf.votedFor = rf.me  // Vote for self
 }
 
+func (rf *Raft) toLeader() {
+	rf.state = LEADER
+	rf.leader = rf.me
+
+	// clear states
+	rf.nextIndex = make([]int, len(rf.peers))
+	rf.matchIndex = make([]int, len(rf.peers))
+
+	// send heartbeat
+	rf.sendHeartbeat()
+}
+
+func (rf *Raft) vote(voteChan chan int, replies []RequestVoteReply) {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+
+	lastLog := getLastLog(rf.log)
+	lastLogIndex, lastLogTerm := -1, -1
+	if (lastLog != nil) {
+		lastLogIndex = lastLog.Index
+		lastLogTerm = lastLog.Term
+	}
+	args := RequestVoteArgs{
+		Term: rf.currentTerm,
+		CandidateId: rf.me,
+		LastLogIndex: lastLogIndex,
+		LastLogTerm: lastLogTerm,
+	}
+
+	for i := range rf.peers {
+		if (i != rf.me) {
+			go rf.sendRequestVote(i, &args, &replies[i])
+			voteChan <- i
+		}
+	}
+}
+
+func (rf *Raft) countVotes(voteChan chan int, replies []RequestVoteReply) {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+
+	votes := 1  // vote for self
+	for _ = range replies {
+		peer := <-voteChan
+		reply := replies[peer]
+
+		// If AppendEntries RPC received from new leader: convert to follower
+		if (reply.Term > rf.currentTerm) {
+			rf.toFollower(reply.Term)
+			return
+		}
+		if (reply.VoteGranted) {
+			votes += 1
+		}
+
+		// If votes received from majority of servers: become leader
+		isMajority := votes > len(replies) / 2
+		if (isMajority && rf.state == CANDIDATE) {
+			rf.toLeader()
+		}
+
+		// If election timeout elapses: start new election?
```

Comment:
You can use time.Timer to do this. Otherwise other way might be a little bit difficult to implement